File: src/main/java/com/kasisoft/libs/common/functional/KPredicate.java
package com.kasisoft.libs.common.functional;

import com.kasisoft.libs.common.*;

import jakarta.validation.constraints.*;

import java.util.function.*;

/**
 * @author daniel.kasmeroglu@kasisoft.com
 */
@FunctionalInterface
public interface KPredicate<T> {

    boolean test(T arg) throws Exception;

    default KPredicate<T> and(@NotNull KPredicate<? super T> other) {
        return $t -> test($t) && other.test($t);
    }

    @NotNull
    default KPredicate<T> negate() {
        return $t -> !test($t);
    }

    @NotNull
    default KPredicate<T> or(@NotNull KPredicate<? super T> other) {
        return $t -> test($t) || other.test($t);
    }

    @NotNull
    default Predicate<T> protect() {
        return $t -> {
            try {
                return test($t);
            } catch (Exception ex) {
                throw KclException.wrap(ex);
            }
        };
    }

} /* ENDINTERFACE */

File: src/main/kotlin/com/kasisoft/libs/common/functional/KPredicate.kt
package com.kasisoft.libs.common.functional;

import com.kasisoft.libs.common.*;

import java.util.function.*;

/**
 * This implementation allows to throw Exceptions (useful on the Java side)
 *
 * @author daniel.kasmeroglu@kasisoft.com
 */
fun interface KPredicate<T: Any?> {

    @Throws(Exception::class)
    fun test(t: T): Boolean

    fun protect(): Predicate<T> =
        Predicate<T> { t: T ->
            try {
                test(t)
            } catch (ex: Exception) {
                throw KclException.wrap(ex)
            }
        }

    fun negate(): Predicate<T> =
        Predicate { t: T ->
            try {
                !test(t)
            } catch (ex: Exception) {
                throw KclException.wrap(ex)
            }
        }

    fun and(after: Predicate<T>) =
        Predicate { t: T ->
            try {
                test(t) && after.test(t)
            } catch (ex: Exception) {
                throw KclException.wrap(ex)
            }
        }

    fun and(after: KPredicate<T>) =
        Predicate { t: T ->
            try {
                test(t) && after.test(t)
            } catch (ex: Exception) {
                throw KclException.wrap(ex)
            }
        }

    fun or(after: Predicate<T>) =
        Predicate { t: T ->
            try {
                test(t) || after.test(t)
            } catch (ex: Exception) {
                throw KclException.wrap(ex)
            }
        }

    fun or(after: KPredicate<T>) =
        Predicate { t: T ->
            try {
                test(t) || after.test(t)
            } catch (ex: Exception) {
                throw KclException.wrap(ex)
            }
        }

} /* ENDINTERFACE */
